pkg/config/remote/uptane/transactional_store.go

```diff
@@ -0,0 +1,227 @@
+	if !ok {
+		cachedBucket = make(dbBucket)
+		ts.cachedData[bucketName] = cachedBucket
+	}
+	return cachedBucket
+}
+
+// Represents a transaction around the store.
+// It doesn't provide any locking, as it's all managed by View/Update calls
+type transaction struct {
+	store *transactionalStore
+}
+
+func (t *transaction) put(bucketName string, path string, data []byte) {
+	bucket := t.store.getMemBucket(bucketName)
+	bucket[path] = data
+}
+
+func (t *transaction) delete(bucketName string, path string) {
+	bucket := t.store.getMemBucket(bucketName)
+	bucket[path] = nil
+}
+
+func (t *transaction) get(bucketName string, path string) ([]byte, error) {
+	bucket := t.store.getMemBucket(bucketName)
+
+	// check if it's present in the in-memory cache
+	data, ok := bucket[path]
+	if ok {
+		return data, nil
+	}
+
+	// fallback to DB access
+	err := t.store.db.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte(bucketName))
+		if bucket == nil {
+			return nil
+		}
+		data = bucket.Get([]byte(path))
+		return nil
+	})
+
+	if len(data) == 0 {
+		err = errors.Wrapf(err, "File empty or not found: %s in bucket %s", path, bucketName)
+	}
+
+	return data, err
+}
+
+type pathData struct {
+	path string
+	data []byte
+}
+
+func (ts *transactionalStore) getAll(bucketName string) ([]pathData, error) {
+	seenBlobs := map[string]struct{}{}
+	blobs := []pathData{}
+
+	for path, data := range ts.getMemBucket(bucketName) {
+		if len(data) == 0 {
+			continue
+		}
+		seenBlobs[path] = struct{}{}
+		blobs = append(blobs, pathData{path, data})
+	}
+
+	err := ts.db.View(func(tx *bbolt.Tx) error {
+		metaBucket := tx.Bucket([]byte(bucketName))
+		if metaBucket == nil {
+			return nil
+		}
+
+		cursor := metaBucket.Cursor()
+		for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+			_, ok := seenBlobs[string(k)]
+			if ok {
+				continue
+			}
+
+			if len(v) == 0 {
+				continue
+			}
+
+			tmp := make([]byte, len(v))
+			copy(tmp, v)
+			blobs = append(blobs, pathData{string(k), tmp})
+		}
+		return nil
+	})
+	return blobs, err
+}
+
+// transaction types
+func (ts *transactionalStore) view(fn func(*transaction) error) error {
+	ts.lock.RLock()
+	defer ts.lock.RUnlock()
+	return fn(&transaction{ts})
+}
+
+func (ts *transactionalStore) update(fn func(*transaction) error) error {
+	ts.lock.Lock()
+	defer ts.lock.Unlock()
+	err := fn(&transaction{ts})
+	return err
+}
+
+func (t *transaction) getAll(bucketName string) ([]pathData, error) {
+	return t.store.getAll(bucketName)
+}
+
+func (t *transaction) pruneTargetFiles(bucketName string, keptPaths []string) {
+	kept := make(map[string]struct{})
+	for _, k := range keptPaths {
+		kept[k] = struct{}{}
+	}
+
+	// delete from in-memory
+	memBucket := t.store.getMemBucket(bucketName)
+	for path := range memBucket {
+		if _, keep := kept[path]; !keep {
+			t.delete(bucketName, path)
+		}
+	}
+
+	// delete in-memory based on files in the DB
+	_ = t.store.db.View(func(tx *bbolt.Tx) error {
```

Comment:
Nothing should happen as it only operates on the database read only and after reading the data it just deletes the remaining targets from memory.
Doesn't hurt to be more safer, it will kill the transaction and force a rollback, but then again if we can't read the database even in read only mode we have bigger fish to fry